store/copr/batch_coprocessor.go

```diff
@@ -798,15 +781,108 @@ func failpointCheckWhichPolicy(act tiflashcompute.DispatchPolicy) {
+	needCrossZoneAccess := false
+	allStores := cache.GetAllValidTiFlashStores(task.region, rpcCtx.Store, tikv.LabelFilterNoTiFlashWriteNode)
+	allStores, needCrossZoneAccess = filterAllStoresAccordingToTiFlashReplicaRead(allStores, aliveStoreIDsInAllZones, aliveStoreIDsInTiDBZone, isTiDBLabelZoneSet, tiflashReplicaReadPolicy)
+	regionInfo = RegionInfo{Region: task.region, Meta: rpcCtx.Meta, Ranges: task.ranges, AllStores: allStores, PartitionIndex: task.partitionIndex}
+	if needCrossZoneAccess && !tiflashReplicaReadPolicy.IsAllReplicas() {
+		regionsInOtherZones = append(regionsInOtherZones, task.region.GetID())
+		regionInfoNeedsReloadOnSendFail = append(regionInfoNeedsReloadOnSendFail, regionInfo)
+		if tiflashReplicaReadPolicy.IsClosestReplicas() && len(regionsInOtherZones) > maxRemoteReadCountAllowed {
+			regionIDErrMsg := ""
+			for i := 0; i < 3 && i < len(regionsInOtherZones); i++ {
+				regionIDErrMsg += fmt.Sprintf("%d, ", regionsInOtherZones[i])
+			}
+			err = errors.Errorf("no less than %d region(s) can not be accessed by TiFlash in the zone [%s]: %setc", len(regionsInOtherZones), tidbZone, regionIDErrMsg)
+			// We need to reload the region cache here to avoid the failure throughout the region cache refresh TTL.
+			cache.OnSendFailForBatchRegions(bo, rpcCtx.Store, regionInfoNeedsReloadOnSendFail, true, err)
+			return regionInfo, nil, nil, err
+		}
+	}
+	return regionInfo, regionInfoNeedsReloadOnSendFail, regionsInOtherZones, nil
+}
+
 // When `partitionIDs != nil`, it means that buildBatchCopTasksCore is constructing a batch cop tasks for PartitionTableScan.
 // At this time, `len(rangesForEachPhysicalTable) == len(partitionIDs)` and `rangesForEachPhysicalTable[i]` is for partition `partitionIDs[i]`.
 // Otherwise, `rangesForEachPhysicalTable[0]` indicates the range for the single physical table.
-func buildBatchCopTasksCore(bo *backoff.Backoffer, store *kvStore, rangesForEachPhysicalTable []*KeyRanges, storeType kv.StoreType, isMPP bool, ttl time.Duration, balanceWithContinuity bool, balanceContinuousRegionCount int64) ([]*batchCopTask, error) {
+func buildBatchCopTasksCore(bo *backoff.Backoffer, store *kvStore, rangesForEachPhysicalTable []*KeyRanges, storeType kv.StoreType, isMPP bool, ttl time.Duration, balanceWithContinuity bool, balanceContinuousRegionCount int64, tiflashReplicaReadPolicy tiflash.ReplicaRead, appendWarning func(error)) ([]*batchCopTask, error) {
 	cache := store.GetRegionCache()
 	start := time.Now()
 	const cmdType = tikvrpc.CmdBatchCop
 	rangesLen := 0
 
+	tidbZone, isTiDBLabelZoneSet := config.GetGlobalConfig().Labels[placement.DCLabelKey]
+	var (
+		aliveStores               []*tikv.Store
+		aliveStoreIDsInAllZones   map[uint64]struct{}
+		aliveStoreIDsInTiDBZone   map[uint64]struct{}
+		maxRemoteReadCountAllowed int
+		careTiFlashReplicaRead    = !tiflashReplicaReadPolicy.IsAllReplicas() && isTiDBLabelZoneSet
```

Comment:
How about just overwrite `tiflashReplicaReadPolicy` to `all_replicas` if TiDBLableZone is not set, so we don't need to handle the conflicts of `tiflashReplicaReadPolicy` and `isTiDBLabelZoneSet` in the following code, and `careTiFlashReplicaRead` can also be removed.